Cater for parsing multiline traps

duncs opened the pull request:
Some traps might have one element spread over multiple lines, such as:

hostname.somedomain
UDP: [192.168.100.10]:61613->[192.168.100.10]:162
DISMAN-EVENT-MIB::sysUpTimeInstance 0:0:00:00.01
SNMPv2-MIB::snmpTrapOID.0 saatrap::snmpSAAEvent
saatrap::saaEventName "Message Sent"
saatrap::saaEventDescription "Message Partner DataStoreIn, Session 2129 - Message sent
    Sequence number : 1
    UUMID           : ABCDABCDABCDABCDABCDABCD123
    Suffix          : 1059726480192
"
SNMP-COMMUNITY-MIB::snmpTrapAddress.0 192.168.100.10

This patch caters for reconstituting the lines where those lines are enclosed
with double quotes.  If only one quote is found from that point to the end of
the trap then the original packet is used, as now.

This patch also has the effect of fixing variable numbers as the multiple
lines are counted as variables in their own right, e.g. in the above $P11
is SNMP-COMMUNITY-MIB::snmpTrapAddress, not the expected $P7

duncs committed:
Cater for parsing multiline traps

Some traps might have one element spread over multiple lines, such as:

hostname.somedomain
UDP: [192.168.100.10]:61613->[192.168.100.10]:162
DISMAN-EVENT-MIB::sysUpTimeInstance 0:0:00:00.01
SNMPv2-MIB::snmpTrapOID.0 saatrap::snmpSAAEvent
saatrap::saaEventName "Message Sent"
saatrap::saaEventDescription "Message Partner DataStoreIn, Session 2129 - Message sent
    Sequence number : 1
    UUMID           : ABCDABCDABCDABCDABCDABCD123
    Suffix          : 1059726480192
"
SNMP-COMMUNITY-MIB::snmpTrapAddress.0 192.168.100.10

This patch caters for reconstituting the lines where those lines are enclosed
with double quotes.  If only one quote is found from that point to the end of
the trap then the original packet is used, as now.

This patch also has the effect of fixing variable numbers as the multiple
lines are counted as variables in their own right, e.g. in the above $P11
is SNMP-COMMUNITY-MIB::snmpTrapAddress, not the expected $P7

duncs committed:
Fix double quoting on expanded trap variables

Give a line in the trap such as
  snmptrap::identifier "value"
and given a variable to expand such as
  my $var = $trap->expand('"${V3}"');
you get the following:
  $var = '""value""'
which $trap->eval will then fall over with

Look for double quoting (using double quotes only) in these situations and
remove one level of quote chars

duncs committed:
Revert "Fix double quoting on expanded trap variables"

This reverts commit f6baf7114f2b83921473b485501708e3a4239ad3.

duncs committed:
Allow simple calculations within expand()

Add in 'eval { ... }' within expand() to allow for simple calculations of
variables performed within a Safe compartment

    my $result = $cp->expand( 'eval{ ${V5} * {$V9} }');
    my $result = $cp->expand( 'eval { sprintf("%.2f", ${V12} / {$V5} ) }');

tonvoon commented on t/SNMP-Trapinfo.t:
```diff
@@ -332,3 +333,52 @@ like( $@, '/Bad ref/', "Complain if bad parameters for new()");
+cmp_ok( $trap->expand('${SNMP-COMMUNITY-MIB::snmpTrapAddress}'), 'eq', '192.168.100.10', "SNMP-COMMUNITY-MIB::snmpTrapAddress is correct on multiline trap");
+cmp_ok( $trap->expand('${SNMPv2-MIB::snmpTrapEnterprise}'), 'eq', 'saatrap::saa', "SNMPv2-MIB::snmpTrapEnterprise is correct on multiline trap");
+cmp_ok( $trap->expand('${V16}'), 'eq', 'saatrap::saa', "V16 is correct on multiline trap");
+ok( $trap->expand('${saatrap::saaEventDescription}') =~ m/ABCDABCDABCDABCDABCDABCD123/, "saatrap::saaEventDescription contains data from multiple lines");
```
Can you add a test that looks for the exact text of this value, rather than just a regexp. This will prove all the linefeeds are retained

tonvoon commented on t/SNMP-Trapinfo.t:
```diff
@@ -1,11 +1,12 @@
 # change 'tests => 1' to 'tests => last_test_to_print';
 
-use Test::More tests => 100;
+use Test::More tests => 114;
```
This can be changed to a done_testing() at the end, to avoid merge conflicts

tonvoon reviewed (commented):
Need to include a doc update, explaining that multi-lines are permissible, so values may contain line feeds

duncs committed:
Improve tests

- Do a full trap comparison for the multi-line test
- Swap to 'done_testing();' to ignreo test count

duncs committed:
Merge remote-tracking branch 'upstream/master' into multiline_traps

duncs committed:
Merge remote-tracking branch 'origin/multiline_traps'

duncs committed:
Docs for eval within expand

duncs committed:
Add in multi-line documentation

Cover values containing multiple lines and ensure double-quoting
is specified.
